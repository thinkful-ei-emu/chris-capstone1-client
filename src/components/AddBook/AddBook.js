import React, { Component } from 'react';
import ListContext from '../../context/ListContext';
import BookApiService from '../../services/book-api-service';
import { Section, Button, Input } from '../Utils/Utils';

export default class AddBookForm extends Component {
    static contextType = ListContext;

    state = {
        error: null,
        title: '',
        author: '',
        list: 'wishlist',
        book_source: 'Print',
        genreff: 'fiction',
        fgenre: 'Action and Adventure',
        nfgenre: 'Art',
        book_report: '',
        image: '',
        rating: 1
    };

    handleSubmit = e => {
        e.preventDefault()
        const { title, author, list, book_source, fgenre, nfgenre,
                book_report, image, rating, genreff } = this.state;
        let genre;
        if(genreff === 'fiction'){
            genre = fgenre;
        } else {
            genre = nfgenre;
        }
        
        const book = { title, author, list, book_source, genre,
            book_report, image, rating };
        
        this.setState({ error: null })
        BookApiService.postBook(book)
        .then(data => {
            this.context.addBook(data)
            window.open('/yourshelf','_self');
        })
        .catch(error => {
            this.setState({ error })
        })
    }

    handleCancelClick = () => this.props.history.push('/yourshelf')
    updateBookTitle = title => this.setState({ title })
    updateBookAuthor = author => this.setState({ author })
    updateBookList = list => this.setState({ list })
    updateBookSource = book_source => this.setState({ book_source })
    updateBookGenreFF = genreff => this.setState({ genreff })
    updateBookGenre = genre => this.setState({ genre })
    updateBookReport = book_report => this.setState({ book_report })
    updateBookImage = image => this.setState({ image })
    updateBookRating = rating => this.setState({ rating })

    renderGenre(){
        const { genreff } = this.state;
        const fictionList = ['Action and Adventure',
        'Alternate history',
        'Anthology',
        'Childrens book',
        'Comic book',
        'Coming-of-age',
        'Crime',
        'Drama',
        'Fairytale',
        'Fantasy',
        'Graphic novel',
        'Historical fiction',
        'Horror',
        'Mystery',
        'Paranormal romance',
        'Picture book',
        'Poetry',
        'Political thriller',
        'Romance',
        'Satire',
        'Science fiction',
        'Short story',	
        'Suspense',
        'Thriller',
        'Young adult'];
        const nonFicitonList = ['Art',
        'Autobiography',
        'Biography',
        'Book review',
        'Cookbook',
          'Diary',
        'Dictionary',
        'Encyclopedia',
        'Guide',
        'Health',
        'History',
        'Journal',
        'Math',
        'Memoir',
        'Prayer',
        'Religion, spirituality, and new age',
        'Textbook',
          'Review',
        'Science',
        'Self help',
        'Travel',
        'True crime']

        let genre;
        let title = '';
        let tag = '';
        let state;
        if(genreff === 'fiction') {
            // this.setState({ genre: 'Action and Adventure' })
            title = 'Fiction Genres:';
            tag = 'fgenre';
            state = this.state.fgenre;
            genre = fictionList.map(gre => <option value={gre}>{gre}</option>)
        } else if(genreff === 'nonfiction') {
            // this.setState({ genre: 'Art' })
            title = 'Non-Fiction Genres:'
            tag = 'nfgenre'
            state = this.state.nfgenre;
            genre = nonFicitonList.map(gre => <option value={gre}>{gre}</option>)
        } else {
            genre = <div></div>
        }
        return (
            <div>
                <label htmlFor={tag}>{title} {' '} </label>
                <select name={tag} id={tag}
                value={state} required
                onChange={e => this.updateBookGenre(e.target.value)}>
                    {genre}
                </select>
            </div>
        );
    }

    renderBookSource(){
        const source = ['Print',
        'Kindle',
        'Nook',
        'Website',
        'Magazine',
        'Journel',
        'Audiobook',
        'Other eBook Source',
        'Other'];
        const book_source = source.map(bs => <option value={bs}>{bs}</option>)
        return book_source;
    }

    render() {
        const { title, author, list, book_source,
            book_report, image, rating, error, genreff } = this.state;
        return(
            <Section className='EditBookForm'>
                <h1>Add Book</h1>
                <form className='EditBook_form'
                    onSubmit={this.handleSubmit}
            >
                <div className='Edit_error' role='alert'>
                {error && <p>{error.message}</p>}
                </div>
                <div>
                    <label htmlFor='title'>Title {' '} </label>
                    <Input type='text' name='title' id='title'
                    value={title} required
                    onChange={e => this.updateBookTitle(e.target.value)}/>
                </div>
                <div>
                    <label htmlFor='author'>Author {' '} </label>
                    <Input type='text' name='author' id='author'
                    value={author} required
                    onChange={e => this.updateBookAuthor(e.target.value)}/>
                </div>
                <div>
                    <label htmlFor='list'>List {' '} </label>
                    <select name='list' id='list'
                    value={list} required
                    onChange={e => this.updateBookList(e.target.value)}>
                        <option value='wishlist'>Wish-List</option>
                        <option value='reading'>Current Reads</option>
                        <option value='read'>Finished Reads</option>
                    </select>
                </div>
                <div>
                    <label htmlFor='book_source'>Book Source {' '} </label>
                    <select name='book_source' id='book_source'
                    value={book_source} required
                    onChange={e => this.updateBookSource(e.target.value)}>
                        {this.renderBookSource()}
                    </select>
                </div>
                <div>
                    <label htmlFor='writing_form'>Writing Form {' '} </label>
                    <select name='writing_form' id='writing_form'
                    value={genreff} required
                    onChange={e => this.updateBookGenreFF(e.target.value)}>
                        <option value='fiction'>Fiction</option>
                        <option value='nonfiction'>Non-Fiction</option>
                    </select>
                </div>
                {this.renderGenre()}
                <div>
                    <label htmlFor='image'>Image {' '} </label>
                    <Input type='url' name='image' id='image'
                    value={image}
                    onChange={e => this.updateBookImage(e.target.value)}/>
                </div>
                <div>
                    <label htmlFor='rating'>Report Score (1-100) {' '} </label>
                    <Input type='number' name='rating' id='rating'
                    value={rating} min='1' max='100'
                    onChange={e => this.updateBookRating(e.target.value)}/>
                </div>
                <div>
                        <label htmlFor='book_report'>
                        BookReport
                        </label>
                        <textarea
                        name='book_report'
                        id='book_report'
                        value={book_report}
                        onChange={e => this.updateBookReport(e.target.value)}
                        />
                    </div>
                <div className='EditBook__buttons'>
                        <Button type='submit'>
                        Add
                        </Button>
                        {' '}
                        <Button type='button' onClick={this.handleCancelClick}>
                        Cancel
                        </Button>
                    </div>
            </form>
            </Section>
        )
    }
}