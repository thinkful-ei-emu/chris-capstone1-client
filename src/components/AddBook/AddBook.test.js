import React from 'react';
import { shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import AddBook from './AddBook';
import ReactDom from 'react-dom';

describe('AddBook component', () => {
    it('renders without crashing', () => {
        const div = document.createElement('div')
        ReactDom.render(
            <AddBook />,
          div
        )
        ReactDom.unmountComponentAtNode(div)
      })
    it('renderes AddBook by default', () => {
        expect(toJson(shallow(<AddBook />))).toMatchSnapshot() 
    })
})