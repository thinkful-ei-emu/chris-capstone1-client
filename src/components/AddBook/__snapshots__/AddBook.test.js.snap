// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AddBook component renderes AddBook by default 1`] = `
<Section
  className="AddBookForm"
>
  <h1>
    Add Book
  </h1>
  <form
    className="AddBook_form"
    onSubmit={[Function]}
  >
    <div
      className="Add_error"
      role="alert"
    />
    <div>
      <label
        htmlFor="title"
      >
        Title 
         
         
      </label>
      <Input
        id="title"
        name="title"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        htmlFor="author"
      >
        Author 
         
         
      </label>
      <Input
        id="author"
        name="author"
        onChange={[Function]}
        required={true}
        type="text"
        value=""
      />
    </div>
    <div>
      <label
        htmlFor="list"
      >
        List 
         
         
      </label>
      <select
        id="list"
        name="list"
        onChange={[Function]}
        required={true}
        value="wishlist"
      >
        <option
          value="wishlist"
        >
          Wish-List
        </option>
        <option
          value="reading"
        >
          Current Reads
        </option>
        <option
          value="read"
        >
          Finished Reads
        </option>
      </select>
    </div>
    <div>
      <label
        htmlFor="book_source"
      >
        Book Source 
         
         
      </label>
      <select
        id="book_source"
        name="book_source"
        onChange={[Function]}
        required={true}
        value="Print"
      >
        <option
          value="Print"
        >
          Print
        </option>
        <option
          value="Kindle"
        >
          Kindle
        </option>
        <option
          value="Nook"
        >
          Nook
        </option>
        <option
          value="Website"
        >
          Website
        </option>
        <option
          value="Magazine"
        >
          Magazine
        </option>
        <option
          value="Journel"
        >
          Journel
        </option>
        <option
          value="Audiobook"
        >
          Audiobook
        </option>
        <option
          value="Other eBook Source"
        >
          Other eBook Source
        </option>
        <option
          value="Other"
        >
          Other
        </option>
      </select>
    </div>
    <div>
      <label
        htmlFor="writing_form"
      >
        Writing Form 
         
         
      </label>
      <select
        id="writing_form"
        name="writing_form"
        onChange={[Function]}
        required={true}
        value="fiction"
      >
        <option
          value="fiction"
        >
          Fiction
        </option>
        <option
          value="nonfiction"
        >
          Non-Fiction
        </option>
      </select>
    </div>
    <div>
      <label
        htmlFor="fgenre"
      >
        Fiction Genres:
         
         
         
      </label>
      <select
        id="fgenre"
        name="fgenre"
        onChange={[Function]}
        required={true}
        value="Action and Adventure"
      >
        <option
          value="Action and Adventure"
        >
          Action and Adventure
        </option>
        <option
          value="Alternate history"
        >
          Alternate history
        </option>
        <option
          value="Anthology"
        >
          Anthology
        </option>
        <option
          value="Childrens book"
        >
          Childrens book
        </option>
        <option
          value="Comic book"
        >
          Comic book
        </option>
        <option
          value="Coming-of-age"
        >
          Coming-of-age
        </option>
        <option
          value="Crime"
        >
          Crime
        </option>
        <option
          value="Drama"
        >
          Drama
        </option>
        <option
          value="Fairytale"
        >
          Fairytale
        </option>
        <option
          value="Fantasy"
        >
          Fantasy
        </option>
        <option
          value="Graphic novel"
        >
          Graphic novel
        </option>
        <option
          value="Historical fiction"
        >
          Historical fiction
        </option>
        <option
          value="Horror"
        >
          Horror
        </option>
        <option
          value="Mystery"
        >
          Mystery
        </option>
        <option
          value="Paranormal romance"
        >
          Paranormal romance
        </option>
        <option
          value="Picture book"
        >
          Picture book
        </option>
        <option
          value="Poetry"
        >
          Poetry
        </option>
        <option
          value="Political thriller"
        >
          Political thriller
        </option>
        <option
          value="Romance"
        >
          Romance
        </option>
        <option
          value="Satire"
        >
          Satire
        </option>
        <option
          value="Science fiction"
        >
          Science fiction
        </option>
        <option
          value="Short story"
        >
          Short story
        </option>
        <option
          value="Suspense"
        >
          Suspense
        </option>
        <option
          value="Thriller"
        >
          Thriller
        </option>
        <option
          value="Young adult"
        >
          Young adult
        </option>
      </select>
    </div>
    <div>
      <label
        htmlFor="image"
      >
        Image Link (must have 'https://') 
         
         
      </label>
      <Input
        id="image"
        name="image"
        onChange={[Function]}
        type="url"
        value=""
      />
    </div>
    <div
      className="AddBook__buttons"
    >
      <Button
        type="submit"
      >
        Add
      </Button>
       
      <Button
        onClick={[Function]}
        type="button"
      >
        Cancel
      </Button>
    </div>
  </form>
</Section>
`;
